persistence/core/src/main/scala/com/lightbend/lagom/internal/spi/projection/ProjectionSpi.scala

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (C) Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package com.lightbend.lagom.internal.spi.projection
+
+import akka.actor.ActorSystem
+import akka.annotation.InternalStableApi
+import akka.persistence.query.EventEnvelope
+import akka.persistence.query.Offset
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+object ProjectionSpi {
+  @InternalStableApi
+  private[lagom] def startProcessing(system: ActorSystem, tagName: String, envelope: EventEnvelope): EventEnvelope =
+    envelope
+
+  @InternalStableApi
+  private[lagom] def afterUserFlow(projectionName: String, offset: Offset): Offset = offset
+
+  @InternalStableApi
+  private[lagom] def completedProcessing(offset: Future[Offset], exCtx: ExecutionContext): Future[Offset] = offset
+
+}
```

Comment:
This is the main change in the PR.

Reply:
```
  private[lagom] def startProcessing(system: ActorSystem, tagName: String, envelope: EventEnvelope): 
```

This is invoked as soon as an event is read from the journal. See `AbstractPersistentnEntityRegistry.eventStream`

```
  private[lagom] def afterUserFlow(projectionName: String, offset: Offset): Offset = offset
```

This is invoked, if possible, immediately after completing the user code in the stream. This is sometimes not possible (depends on the projection implementation).

```
  private[lagom] def completedProcessing(offset: Future[Offset], exCtx: ExecutionContext): Future[Offset] 
```

This is invoked when the offset has been persisted.

```
  private[lagom] def failed(actorSystem: ActorSystem, projectionName: String, partitionName: String): ```

This is invoked is a failure traverses the stream on the porjection.